src/coreclr/jit/morph.cpp

```diff
@@ -4361,19 +4564,34 @@ GenTree* Compiler::fgMorphMultiregStructArg(GenTree* arg, fgArgTabEntry* fgEntry
     {
         assert(structSize <= MAX_ARG_REG_COUNT * TARGET_POINTER_SIZE);
         BYTE gcPtrs[MAX_ARG_REG_COUNT];
-        elemCount = roundUp(structSize, TARGET_POINTER_SIZE) / TARGET_POINTER_SIZE;
         info.compCompHnd->getClassGClayout(objClass, &gcPtrs[0]);
+        elemCount = roundUp(structSize, TARGET_POINTER_SIZE) / TARGET_POINTER_SIZE;
+#ifdef TARGET_LOONGARCH64
+        // For LoongArch64's ABI, the struct which size is TARGET_POINTER_SIZE
+        // may be passed by two registers.
+        // e.g  `struct {int a; float b;}` passed by an integer register and a float register.
+        if (fgEntryPtr->numRegs == 2)
+        {
+            elemCount = 2;
+        }
+#endif
 
         for (unsigned inx = 0; inx < elemCount; inx++)
         {
-#ifdef UNIX_AMD64_ABI
+#if defined(UNIX_AMD64_ABI)
             if (gcPtrs[inx] == TYPE_GC_NONE)
             {
                 type[inx] = GetTypeFromClassificationAndSizes(fgEntryPtr->structDesc.eightByteClassifications[inx],
                                                               fgEntryPtr->structDesc.eightByteSizes[inx]);
             }
             else
-#endif // UNIX_AMD64_ABI
+#elif defined(TARGET_LOONGARCH64)
+            if (fgEntryPtr->structFloatFieldType[inx] != TYP_UNDEF)
```

Comment:
Nice, this is much better 👍.

Two concerns: first, `struct { int, double }`/`struct { float, double }` will probably have their second field loaded at offset `4` instead of `8` (same issue as for the return descriptor). Long-term we should refactor this method to track `union { type, argSize }[MAX_REG_COUNT]`, but for LA here it should suffice to "fix up" the first field's type to be at least as large as the second (same as the return descriptor code does) after the loop.

Second, there is this code below:
```
// We setup the type[inx] value above using the GC info from 'objClass'
// This GT_LCL_VAR must have the same GC layout info
//
if (varTypeIsGC(currentGcLayoutType))
{
    noway_assert(type[inx] == currentGcLayoutType);
}
else
{
    // We may have use a small type when we setup the type[inx] values above
    // We can safely widen this to TYP_I_IMPL
    type[inx] = TYP_I_IMPL;
}
```
Which I suspect will try to turn `{ float, float }` into `{ long, long }`.

Reply:
Yeah, I had updated within the https://github.com/dotnet/runtime/pull/62843/commits/93e27c02ee37e088abe992bf2d19fd9f5c3f062c